Fix some Cos-scaling settings

jm-c opened the pull request:
## What changes does this PR introduce?
Implement changes as suggested in issue #851

## What is the current behaviour? 
- inconsistent ISOTROPIC_COS_SCALING code in `mom_vecinv/mom_vi_hdissip.F` 
- potentially confusing setting of cos-scaling CPP options in `GAD_OPTIONS.h` (for tracer diffusivity) and in `CPP_OPTIONSh` for remaining code as `CPP_OPTIONS.h` is also included in all `pkg/generic_advdiff` src files.

## What is the new behaviour 
- remove all ISOTROPIC_COS_SCALING code from  `mom_vecinv/mom_vi_hdissip.F`
- clarify  cos-scaling CPP option setting by moving them from  `CPP_OPTIONS.h`  to `MOM_COMMON_OPTIONS.h` to be used for horizontal viscosity; and keep second setting in `GAD_OPTIONS.h` for horizontal diffusivity.

## Does this PR introduce a breaking change? 
no

## Other information:

## Suggested addition to `tag-index`
o pkg/mom_common,fluxform,vecinv:
  - remove ISOTROPIC_COS_SCALING code from pkg/mom_vecinv/mom_vi_hdissip.F (was
    not right); add warning (or stop if side-drag) if used with pkg/mom_vecinv.
  - move 2 Cos-Scaling CPP options from CPP_OPTIONS.h to MOM_COMMON_OPTIONS.h
    to be used in horizontal viscosity formulation; more consistent use of
    GAD_OPTIONS.h where these 2 CPP options are used for horizontal diffusion.
  - update all verification exp. customized version of CPP_OPTIONS.h,
    GAD_OPTIONS.h and MOM_COMMON_OPTIONS.h.

mjlosch commented on pkg/generic_advdiff/GAD_OPTIONS.h:
```diff
@@ -19,18 +19,16 @@ C Use this file for selecting options within the GAD package
-C *only for use on a lat-lon grid*
-C Setting this flag here only affects the bi-harmonic tracer terms; to
-C use COSINEMETH_III in the momentum equations set it CPP_OPTIONS.h
+C This flag selects the form of COSINE(lat) scaling of horizontal
+C bi-harmonic diffusivity. *only for use on a lat-lon grid*
+C Setting this flag here only affects tracer diffusivity; to use it
+C in the momentum equations it needs to be set in MOM_COMMON_OPTIONS.h
```
I find this confusion, because first it is bi-harmonic diffusivity and in the next line it is just diffusivity. I would also leas the "only for use ..." in a separate line.
```suggestion
C *only for use on a lat-lon grid*
C This flag selects the form of COSINE(lat) scaling of horizontal
C bi-harmonic diffusivity. 
C Setting this flag here only affects tracer (bi-harmonic) diffusivity; to use it
C in the momentum equations it needs to be set in MOM_COMMON_OPTIONS.h
```

jm-c replied:
I have the impression that the risk of confusion is minimal here: tracer biharmonic diffusion is part of tracer diffusion, so it's not wrong. Here the main point is to stress that momentum is not affected (although the CPP option has the same name); and we don't repeat/mention neither other restriction (lat-lon grid,  cosPower >0, ...).

mjlosch commented on pkg/generic_advdiff/GAD_OPTIONS.h:
```diff
@@ -19,18 +19,16 @@ C Use this file for selecting options within the GAD package
-C use ISOTROPIC_COS_SCALING of the horizontal viscosity terms in the
-C momentum equations set it CPP_OPTIONS.h; the following line
-C even overrides setting the flag in CPP_OPTIONS.h
+C This selects isotropic scaling of horizontal harmonic and bi-harmonic
+C diffusivity when using the COSINE(lat) scaling. *only for lat-lon grid*
+C Setting this flag here only affects tracer diffusivity; to use it
+C in the momentum equations it needs to be set in MOM_COMMON_OPTIONS.h
```
similar to previous comment/suggestion
```suggestion
C *only for lat-lon grid*
C This selects isotropic scaling of horizontal harmonic and bi-harmonic
C diffusivity when using the COSINE(lat) scaling. 
C Setting this flag here only affects tracer diffusivity; to use it
C in the momentum equations it needs to be set in MOM_COMMON_OPTIONS.h
```

jm-c replied:
I like when the restriction "only for lat-lon grid" is part of the main description, on same line, But thinking of this,
I would rather keep it on same line but change it to "-- only on lat-lon grid." (without "*").

mjlosch commented on pkg/mom_common/MOM_COMMON_OPTIONS.h:
```diff
@@ -9,6 +9,18 @@ C Use this file for selecting CPP options within the mom_common package
 #ifdef ALLOW_MOM_COMMON
 C     Package-specific options go here
 
+C This flag selects the form of COSINE(lat) scaling of horizontal
+C bi-harmonic viscosity. *only for use on a lat-lon grid*
+C Setting this flag here only affects momentum viscosity; to use it
+C in the tracer equations it needs to be set in GAD_OPTIONS.h
```
```suggestion
C *only for use on a lat-lon grid*
C This flag selects the form of COSINE(lat) scaling of horizontal
C bi-harmonic viscosity.
C Setting this flag here only affects momentum (bi-harmonic) viscosity;
C to use it in the tracer equations it needs to be set in GAD_OPTIONS.h
```

mjlosch commented on pkg/mom_common/MOM_COMMON_OPTIONS.h:
```diff
@@ -9,6 +9,18 @@ C Use this file for selecting CPP options within the mom_common package
+C in the tracer equations it needs to be set in GAD_OPTIONS.h
+#define COSINEMETH_III
+
+C This selects isotropic scaling of horizontal harmonic and bi-harmonic
+C diffusivity when using the COSINE(lat) scaling. *only for lat-lon grid*
```
```suggestion
C *only for lat-lon grid*
C This selects isotropic scaling of horizontal harmonic and bi-harmonic
C diffusivity when using the COSINE(lat) scaling.
```

jrscott replied:
you mean viscosity here?

jm-c replied:
yes, I am going to fix this.

mjlosch reviewed (approved):
Looks good. I have a few inline suggestions about the comments in the header files. These are suggestions and can be ignored.

One thing: Now each and every verification experiment has a CPP_OPTIONS.h file which still contains these flag (all of the commented/undefined). Do you see that as a problem?

jm-c commented:
I've just added some warning or stop in `config_check.F` related to these 2 CPP-options if using `pkg/mom_vecinv`.

@mjlosch I agree that it would be better to update the set of `CPP_OPTIONS.h` (25 files) in verification exp., and also the few `GAD_OPTIONS.h` and `MOM_COMMON_OPTIONS.h` there (12 more files). I am going to look at this.

jm-c commented:
Since there are few customized CPP_OPTIONS.h, and in order to save time, I suggest to remove completely the old and commented out ECCO_CPPOPTIONS.h lines near the bottom of this file: The last exp. to use `ECCO_CPPOPTIONS.h` was converted in March 2022 (PR #606) so I think it's time to remove this from `model/inc/CPP_OPTONS.h`. @mjlosch do you agree ?

mjlosch commented:
Totally, there is no `ECCO_CPPOPTIONS.h` anymore, so let's remove this commented code!

jm-c commented:
I ended up updating all customized `CPP_OPTIONS.h` in verification/*/code*/ dirs, since few of them were lagging behind (from spring 2004). I kept the same set of defined Options (except few `ALLOW_SRCG` that are now defined), so no changes in compiled code nor output.
Also updated all customized `GAD_OPTIONS.h` and `MOM_COMMON_OPATIONS.h`, and just to make sure to test that code compiles,  I switched COSINEMETH_III & ISOTROPIC_COS_SCALING to #undef & #define respectively, in: `advect_xy/code/GAD_OPTIONS.h` and in: `MLAdjust/code/MOM_COMMON_OPTIONS.h`.

jm-c commented:
I've just updated the suggested addition to `tag-index` in this PR description. But since there are many more updated files as earlier, I would like to ask @mjlosch for a second review.

jm-c commented:
I am going to make an other little adjustment to comments in `CPP_OPTIONS.h`,  related to "SHORTWAVE_HEATING",
replacing "external_forcing.F" with "apply_forcing.F", and removing the second line:
> C Note: this should be a run-time option

since it does not describe anything (+ might change soon, e.g. in the context of PR #750, if we add such run-time switch).
So @mjlosch no need to check again this PR now.

jm-c commented:
No more changes from my side, ready for second review, and @jrscott will update the 3 option tables in the docs.

jm-c commented:
@jrscott I fixed the typo you found. Do you "approve" this PR ? And @mjlosch if you could review again (since I updated many files since your previous review) and assuming this is OK I could think of merging this PR soon after.

jrscott commented:
LGTM

mjlosch reviewed (commented):
Changes look good. I have 2 comments:
1. Now very often:
```
C o Use "OLD" UV discretisation near boundaries (*not* recommended)
C   Note - only works with pkg/mom_fluxform and "no_slip_sides=.FALSE."
C          because the old code did not have no-slip BCs
#undef OLD_ADV_BCS
```
is not in the customised version. Intentional?
2. I think `natl_box` and `global_with_exf` were not updated. Intentional?

jm-c commented:
@mjlosch you are right, I did not import (on purpose) the 2 retired options "OLD_ADV_BCS" and "USE_OLD_EXTERNAL_FORCING" since they will have to be removed at some point. Regarding OLD_GRID_IO, it's a bit different since it still useful in some cases.
And `natl_box` and `global_with_exf` version will be removed in PR #885, so not need to update these ones (+ it simplifies merging/updating PR #885 once this one is merged).

mjlosch commented:
Makes a lot of sense! Go ahead and merge, as far as I am concerned.

jm-c commented:
Thanks @mjlosch and @jrscott I will merge this PR in the coming days (waiting for all daily tests to come back after the new exp. "atm_gray" was added ; plus I might need to update the local `CPP_OPTIONS.h` there).
